github/VisualStudio
# Be consistent about how we use relative paths and Git paths

In versions of `LibGit2Sharp` we were using prior to #2142, `LibGit2Sharp` didn't care whether paths passed contained `/` or `\` and could cope with absolute as well as working directory relative paths. In recent versions of `LibGit2Sharp`, Git paths must be working directory relative and use `/` as their directory separator. If they're not in this format the operation will silently fail (as if the file simply doesn't exist).

Because GitHub for Visual Studio was developed when `LibGit2Sharp` didn't care about directory separators, we often mixed up relative paths and Git paths (which use '/'). This has created the potential for a whole category of bugs and has made the code difficult to reason about. This PR is attempt to bring a bit of sanity to the situation.

### What this PR does

- Use `relativePath` when a parameter contains a Windows style relative path
- Use `gitPath` when a parameter contains a Git style relative path
- Use `Guard.ArgumentIsRelativePath` to check we're not passing absolute paths
- Automatically convert Windows paths to Git style relative paths before using them
- Factor out conversions into `Paths.ToGitPath` and `Paths.ToWindowsPath`
- Fix #2380 which was caused by passing `AddBufferTag` an absolute path
- Fix issue where `PullRequestSessionFile` used a Git relative path and `PullRequestSessionLiveFile` a Windows relative path
- Fix issue where `GetOldFileName` was using a Windows style relative path

### How to test

Here is how to test one of the bugs it fixes

1. Open a solution that contains pull requests
2. Open the pull request detail view without checking out the PR
3. Right-click on a PR file that has comments and `View File`
4. Check that comments appear on comment margin

Fixes #2380
